Restore/recreate corrupted SQLite file of local storage

pgorszkowski-igalia opened the pull request:
#### d5346d022b5cd2324b2540c581baaf194712e5eb
<pre>
Restore/recreate corrupted SQLite file of local storage
<a href="https://bugs.webkit.org/show_bug.cgi?id=250470">https://bugs.webkit.org/show_bug.cgi?id=250470</a>

Reviewed by Sihui Liu.

The current implementation in case of corrupted database file of the
localstorage requires few restart to make it works again.
This fix solves it and if the database file of the localstorage is
currupted it deletes broken file.

* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
(WebKit::SQLiteStorageArea::prepareDatabase):
(WebKit::SQLiteStorageArea::getItemFromDatabase):
(WebKit::SQLiteStorageArea::allItems):
(WebKit::SQLiteStorageArea::setItem):
(WebKit::SQLiteStorageArea::removeItem):
(WebKit::SQLiteStorageArea::clear):
(WebKit::SQLiteStorageArea::handleDatabaseCorruption):
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h:

Canonical link: <a href="https://commits.webkit.org/259573@main">https://commits.webkit.org/259573@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/49ec87807d9f336be7f02c78c94ab13bcfbdbb68

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105058 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14137 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37948 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114321 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174518 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108961 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15276 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5059 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97377 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113334 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110814 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11809 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94810 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39313 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93677 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26437 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80978 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7475 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27796 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7569 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4380 "Found 1 new test failure: fast/images/avif-heif-container-as-image.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13624 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47349 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6595 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9358 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

szewai reviewed (commented):
Some thoughts:
1. We cannot recreate database from cache, because cache does not keep item that is bigger than maximumSizeForValuesKeptInMemory. If we leave out those items, the new database might be inconsistent with previous version. Since corruption should be a rare case, I think it's fine to just start fresh with corrupt error (e.g. close db and deleting files).
2. We need loggings at the places where database files are deleted
3. We need a test to ensure the behavior is expected. It can be an API test doing something like: 
(1) making a malformed file (e.g. an empty file with name "localstorage") in LocalStorage directory
(2) creating a web view and loading a web page to store an item in localStorage
(3) creating a second web view and loading a web page with the same url to read the item from localStorage (to ensure the item is stored successfully)

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -2053,6 +2093,7 @@ void beforeAll()
     WebViewTest::add("WebKitWebView", "default-content-security-policy", testWebViewDefaultContentSecurityPolicy);
     WebViewTest::add("WebKitWebView", "web-extension-mode", testWebViewWebExtensionMode);
     WebViewTest::add("WebKitWebView", "disable-web-security", testWebViewDisableWebSecurity);
+    WebViewTest::add("WebKitWebView", "handle-corrupted-local-storage", testWebViewHandleCorruptedLocalStorage);
```
So this doesn't really fit amazingly well into any existing tests, but I think I would shoehorn it into TestWebsiteData.cpp instead of here.

szewai commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -405,5 +442,18 @@ void SQLiteStorageArea::handleLowMemoryWarning()
+    WebCore::SQLiteFileSystem::deleteDatabaseFile(m_path);
+
+    // Try to recreate the database.
+    prepareDatabase(ShouldCreateIfNotExists::Yes);
```
I think we don't need `prepareDatabase` here because:
1. we are not retrying the failed operation on the new database; we will just return error, 
2. if client does not do any operation after the failed operation (that has led to database deletion), we don't want to have database open; if client performs new operation, `prepareDatabase` will be invoked on that operation.

szewai commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -405,5 +442,15 @@ void SQLiteStorageArea::handleLowMemoryWarning()
+{
+    ASSERT(!isMainRunLoop());
+    if (m_database->isOpen())
+        m_database->close();
```
~SQLiteDatabase() should close database, so this is unnecessary.

szewai commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -162,8 +162,18 @@ bool SQLiteStorageArea::prepareDatabase(ShouldCreateIfNotExists shouldCreateIfNo
+        } else {
+            m_database = nullptr;
+            return false;
+        }
```
With `handleDatabaseCorruptionIfNeeded`, we could change this to:
```
bool openResult  = m_database->open(m_path);
if (!openResult && handleDatabaseCorruptionIfNeeded(m_database->lastError())
    openResult = m_database->open(m_path);

if (!openResult) {
    RELEASE_LOG_ERROR(Storage, "SQLiteStorageArea::prepareDatabase failed to open database at '%s'", m_path.utf8().data());
    m_database = nullptr;
    return false;
}
```

pgorszkowski-igalia replied:
great optimization, thanks

szewai commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -405,5 +442,15 @@ void SQLiteStorageArea::handleLowMemoryWarning()
         m_database->releaseMemory();
 }
 
+void SQLiteStorageArea::handleDatabaseCorruption()
```
Since we are checking the error in all the places, maybe we can change this to something like:
```
bool SQLiteStorageArea::handleDatabaseCorruptionIfNeeded(int databaseError)
{
    if (databaseError != SQLITE_CORRUPT && databaseError != SQLITE_NOTADB)
        return false;

    m_database = nullptr;
    RELEASE_LOG(Storage, "SQLiteStorageArea::handleDatabaseCorruption deletes corrupted database file '%s'", m_path.utf8().data());
    webcore::SQLiteFileSystem::deleteDatabaseFile(m_path);
    return true;
}
```
and use handleDatabaseCorruptionIfNeeded(result) in all the places.

szewai reviewed (commented):
Overall looks good to me

szewai reviewed (approved):
LGTM

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebsiteData.cpp:
```diff
@@ -863,6 +863,45 @@ g_assert_nonnull(data);
+    const char html[] = "<html><script>let foo = (window.localStorage.length ? window.localStorage.getItem('item'):''); window.localStorage.setItem('item','value');</script></html>";
+    auto waitForFooChanged = [&test](WebKitWebView* webView) {
+        char* fooValue;
+            GUniqueOutPtr<GError> error;
```
Careful, indentation is wrong here

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebsiteData.cpp:
```diff
@@ -863,6 +863,45 @@ g_assert_nonnull(data);
+            g_assert_no_error(error.get());
+            jscvalue = webkit_javascript_result_get_js_value(result);
+            fooValue = jsc_value_to_string(jscvalue);
+        return fooValue;
```
This should return a `GUniquePtr<char>` so you don't have to manually `g_free` the result.

pgorszkowski-igalia committed:
Restore/recreate corrupted SQLite file of local storage
https://bugs.webkit.org/show_bug.cgi?id=250470

Reviewed by Sihui Liu.

The current implementation in case of corrupted database file of the
localstorage requires few restart to make it works again.
This fix solves it and if the database file of the localstorage is
currupted it deletes broken file.

* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
(WebKit::SQLiteStorageArea::prepareDatabase):
(WebKit::SQLiteStorageArea::getItemFromDatabase):
(WebKit::SQLiteStorageArea::allItems):
(WebKit::SQLiteStorageArea::setItem):
(WebKit::SQLiteStorageArea::removeItem):
(WebKit::SQLiteStorageArea::clear):
(WebKit::SQLiteStorageArea::handleDatabaseCorruption):
* Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.h:

Canonical link: https://commits.webkit.org/259573@main
